File: create_token.rb
require 'json'
require 'net/http'
require 'uri'
require 'jwt'
require 'base64'
require 'time'

class AppStoreConnectAPI
  def initialize(key_id, issuer_id)
    @key_id = key_id
    @issuer_id = issuer_id
    @private_key_path = "AuthKey_#{key_id}.p8"
  end

  def generate_jwt
    private_key = OpenSSL::PKey::EC.new(File.read(File.expand_path(@private_key_path, __dir__)))

    payload = {
      iss: @issuer_id,
      iat: Time.now.to_i,
      exp: Time.now.to_i + 1200, # 20分钟过期
      aud: 'appstoreconnect-v1'
    }

    JWT.encode(payload, private_key, 'ES256', { kid: @key_id })
  end
 
end

# 使用示例
def main()
  # 从JSON文件读取配置
  config ||= load_config_from_json('iap_config.json')

  # 初始化API客户端
  api = AppStoreConnectAPI.new(config[:key_id], config[:issuer_id])

  puts api.generate_jwt() 
end

def load_config_from_json(file_path)
  config_path = File.expand_path(file_path, __dir__)
  unless File.exist?(config_path)
    raise "配置文件不存在: #{config_path}"
  end

  JSON.parse(File.read(config_path), symbolize_names: true)
end

# 如果直接运行此脚本
if __FILE__ == $0
  puts "App Store Connect API - Token 创建工具"
  puts "=============================================="
  puts ""
  puts "使用前请确保："
  puts "1. 填写 config.json 配置文件里的 key_id 和 issuer_id "
  puts "2. 将私钥文件(.p8)放在同目录下"
  puts ""

  main()
end

File: create_token.swift
import Foundation
import CryptoKit

class AppStoreConnectAPI {
    private let keyID: String
    private let issuerID: String
    private let privateKeyPath: String
    private let baseURL = "https://api.appstoreconnect.apple.com"

    init(keyID: String, issuerID: String) {
        self.keyID = keyID
        self.issuerID = issuerID
        self.privateKeyPath = "AuthKey_\(keyID).p8"
    }

    func generateJWT() throws -> String {
        // Read private key
        let privateKeyURL = URL(fileURLWithPath: privateKeyPath)
        let privateKeyData = try Data(contentsOf: privateKeyURL)

        // Create private key from data
        guard let privateKeyString = String(data: privateKeyData, encoding: .utf8) else {
            throw NSError(domain: "AppStoreConnectAPI", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to read private key as UTF-8 string"])
        }
        let privateKey = try P256.KeyAgreement.PrivateKey(pemRepresentation: privateKeyString)

        // Create JWT header
        let header: [String: Any] = [
//            "alg": "ES256",
            "kid": keyID,
//            "typ": "JWT"
        ]

        // Create JWT payload
        let now = Date()
        let expirationDate = now.addingTimeInterval(1200) // 20 minutes

        let payload: [String: Any] = [
            "iss": issuerID,
            "iat": Int64(now.timeIntervalSince1970),
            "exp": Int64(expirationDate.timeIntervalSince1970),
            "aud": "appstoreconnect-v1"
        ]

        // Encode header and payload
        let headerData = try JSONSerialization.data(withJSONObject: header)
        let payloadData = try JSONSerialization.data(withJSONObject: payload)

        let headerBase64 = headerData.base64EncodedString()
            .replacingOccurrences(of: "=", with: "")
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")

        let payloadBase64 = payloadData.base64EncodedString()
            .replacingOccurrences(of: "=", with: "")
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")

        // Create message to sign
        let message = "\(headerBase64).\(payloadBase64)"
        let messageData = message.data(using: .utf8)!

        // Sign message
        let privateKeyDataForSigning = privateKey.rawRepresentation
        let signingKey = try P256.Signing.PrivateKey(rawRepresentation: privateKeyDataForSigning)
        let signature = try signingKey.signature(for: SHA256.hash(data: messageData))
        let signatureData = signature.rawRepresentation

        let signatureBase64 = signatureData.base64EncodedString()
            .replacingOccurrences(of: "=", with: "")
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")

        // Create final JWT
        return "\(message).\(signatureBase64)"
    }
}

func main() {
    do {
        // You need to provide your actual key ID and issuer ID
        let keyID = "YOUR_KEY_ID"
        let issuerID = "YOUR_ISSUER_ID"

        let api = AppStoreConnectAPI(keyID: keyID, issuerID: issuerID)
        let token = try api.generateJWT()

        print("✅ JWT Token generated successfully!")
        print("Token: \(token)")

        // Print token info
        let tokenParts = token.split(separator: ".")
        print("Header length: \(tokenParts[0].count)")
        print("Payload length: \(tokenParts[1].count)")
        print("Signature length: \(tokenParts[2].count)")
    } catch {
        print("❌ Error generating JWT: \(error)")
    }
}

main()

